script/tool/lib/src/custom_test_command.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file/file.dart';
+import 'package:platform/platform.dart';
+
+import 'common/package_looping_command.dart';
+import 'common/process_runner.dart';
+import 'common/repository_package.dart';
+
+const String _scriptName = 'run_tests.dart';
+const String _legacyScriptName = 'run_tests.sh';
+
+/// A command to run custom, package-local tests on packages.
+///
+/// This is an escape hatch for adding tests that this tooling doesn't support.
+/// It should be used sparingly; prefer instead to add functionality to this
+/// tooling to eliminate the need for bespoke tests.
+class CustomTestCommand extends PackageLoopingCommand {
+  /// Creates a custom test command instance.
+  CustomTestCommand(
+    Directory packagesDir, {
+    ProcessRunner processRunner = const ProcessRunner(),
+    Platform platform = const LocalPlatform(),
+  }) : super(packagesDir, processRunner: processRunner, platform: platform);
+
+  @override
+  final String name = 'custom-test';
+
+  @override
+  final String description = 'Runs package-specific custom tests defined in '
+      'a package\'s tool/$_scriptName file.\n\n'
+      'This command requires "dart" to be in your path.';
+
+  @override
+  Future<PackageResult> runForPackage(RepositoryPackage package) async {
+    final File script =
+        package.directory.childDirectory('tool').childFile(_scriptName);
+    final File legacyScript = package.directory.childFile(_legacyScriptName);
+    String? customSkipReason;
+    bool ranTests = false;
+
+    // Run the custom Dart script if presest.
+    if (script.existsSync()) {
+      final int exitCode = await processRunner.runAndStream(
+          'dart', <String>['run', 'tool/$_scriptName'],
+          workingDir: package.directory);
+      if (exitCode != 0) {
+        return PackageResult.fail();
+      }
+      ranTests = true;
+    }
+
+    // Run the legacy script if present.
+    if (legacyScript.existsSync()) {
```

Comment:
This will make it so we run the dart script and the sh script.  That will cause certain tests to get executed twice since the shell script delegates to the dart script.  We should execute one or the other.

We could juggle things around in Pigeon so that the correct way is to execute the dart script and it will delegate to a shell script test that haven't been migrated (if it isn't windows).  We could even make run_tests.sh just execute the dart script and there is another bash script the dart script delegates to for unmigrated tests.

Reply:
> This will make it so we run the dart script and the sh script.

That's [already how it works](https://github.com/flutter/packages/blob/main/script/local_tests.sh#L62-L85). It had to work that way in order to support pigeon's tests, which were partially implemented in one and partially implemented in the other.

> That will cause certain tests to get executed twice since the shell script delegates to the dart script.

[This code](https://github.com/flutter/packages/blame/main/packages/pigeon/bin/run_tests.dart#L311-L319) was supposed to prevent that. If Pigeon's test structure changed after that in a way that causes it to double-run tests, that's a Pigeon test bug.

> We could juggle things around in Pigeon so that the correct way is to execute the dart script and it will delegate to a shell script test that haven't been migrated (if it isn't windows).

I don't have any opinion on how Pigeon manages the Dart vs shell split in the interim period until all tests are migrated, but it's orthogonal to this PR since the behavior isn't changing.

Reply:
> That's already how it works.

Fair enough.

> that's a Pigeon test bug.

I'm just trying to think what the fix is though given this logic and making sure we have a path forward.  The easiest thing for pigeon would be to rename the script 'run_tests.sh' to 'shell_tests.sh' and make `run_tests.dart` execute `shell_tests.sh`.  If pigeon isn't using `run_tests.sh` then probably no one is so we can get rid of this.

Reply:
> I'm just trying to think what the fix is though given this logic and making sure we have a path forward.

I'm not clear on what the problem actually is. What is the flow where the code I linked to avoids duplicate tests?

> If pigeon isn't using `run_tests.sh` then probably no one is so we can get rid of this.

Four packages currently use `run_test.sh`. See the linked issue for details.